sympy/polys/matrices/domainmatrix.py

```diff
@@ -209,6 +208,54 @@ def from_list_sympy(cls, nrows, ncols, rows, **kwargs):
 
         return DomainMatrix(domain_rows, (nrows, ncols), domain)
 
+    @classmethod
+    def from_dict_sympy(cls, nrows, ncols, elemsdict, **kwargs):
+        """
+
+        Parameters
+        ==========
+
+        nrows: number of rows
+        ncols: number of cols
+        elemsdict: dict of dicts containing non-zero elements of the DomainMatrix
+
+        Returns
+        =======
+
+        DomainMatrix containing elements of elemsdict
+
+        Examples
+        ========
+
+        >>> from sympy.polys.matrices import DomainMatrix
+        >>> from sympy import QQ, ZZ
+        >>> elemsdict = {0: {1: QQ(1, 2)}, 1: {0: ZZ(1)}}
+        >>> DomainMatrix.from_dict_sympy(2, 3, elemsdict)
+        DomainMatrix({0: {1: 1/2}, 1: {0: 1}}, (2, 3), QQ)
+
+        See Also
+        ========
+
+        from_list_sympy
+
+        """
+        if not all(0 <= r < nrows for r in elemsdict):
+            raise DDMBadInputError("Row out of range")
+        if not all(0 <= c < ncols for row in elemsdict.values() for c in row):
+            raise DDMBadInputError("Column out of range")
+
+        items_sympy = [_sympify(item) for row in elemsdict.values() for item in row.values()]
+        domain, items_domain = cls.get_domain(items_sympy, **kwargs)
+
+        idx = 0
+
+        for i, row in zip(elemsdict.keys(), elemsdict.values()):
+                for j in row.keys():
+                    elemsdict[i][j] = items_domain[idx]
```

Comment:
Don't modify the input dict.

Reply:
Ok, I thought we'd have to do it if for eg: our Matrix contains elements `QQ(1, 2), ZZ(1)` so the domain would have to be `QQ`, therefore `ZZ(1)` needs to be changed to `QQ(1)`.

Reply:
This is called `from_dict_sympy` because the input should be `Expr`.

A new dict should be created rather than modifying the dict passed in by the caller.

Reply:
Oh ok, I thought you meant just don't modify elements. My bad, I interpreted that wrong.